### PR TITLE
apps: main backend integration between metadata and chat

### PR DESCRIPTION
**Pull request by loganallenc**
Please don’t merge until I’ve finished the %chat-js new chat page (uses a modified type in the %chat-view API). Otherwise, done! Please review.

Edit: JS changes now attached. Full integration complete, just needs strenuous testing from here on out :)

### CODE REVIEW

File: pkg/arvo/lib/metadata-json.hoon [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  %+  turn  ~(tap by associations)
+  |=  [[=group-path =resource] =metadata]
+  ^-  [cord json]
+  :_  (metadata-to-json metadata)
+  %-  crip
+  ;:  weld
+      (trip (spat group-path))
+      (trip app-name.resource)
+      (trip (spat app-path.resource))
+  ==
```
Comment (Fang-) [inline pkg/arvo/lib/metadata-json.hoon]: This is going to look like `/group-pathapp-name/app-path`.

Even with the `app-name` more cleanly separated from the `group-path`, are we sure it's fine to just concat them into a string like this? I don't think you have any guaranteed on where one part ends and the other begins, if you're a client trying to parse these out.

If these are only going to be used in contexts where you already know the `group-path`, `app-name` and `app-path`, and you only care about doing lookup for those (that you know) specifically, then maybe this is fine... But it's certainly weird!

File: pkg/arvo/lib/metadata-json.hoon [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+++  metadata-to-json
+  |=  met=metadata
+  ^-  json
+  =,  enjs:format
+  %-  pairs
+  :~  [%title s+title.met]
+      [%description s+description.met]
+      [%color s+(scot %ux color.met)]
+      [%date-created s+(scot %da date-created.met)]
+      [%creator s+(scot %p creator.met)]
```
Comment (Fang-) [inline pkg/arvo/lib/metadata-json.hoon]: Probably don't want to bother the frontend with urbit-style data notation. We have dedicated json transformers for `@da` and `@p`.

```hoon
> ((x-co:co 6) 0xff)
"0000ff"
```

https://github.com/urbit/urbit/blob/686aaa565eddfc0311d2e0f43564d4355f4173ec/pkg/arvo/sys/zuse.hoon#L5590-L5592

https://github.com/urbit/urbit/blob/686aaa565eddfc0311d2e0f43564d4355f4173ec/pkg/arvo/sys/zuse.hoon#L5575-L5577

File: pkg/arvo/lib/metadata-json.hoon [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+          [%metadata (metadata-to-json metadata.upd)]
+      ==
+    ?:  ?=(%remove -.upd)
+      :-  %remove
+      %-  pairs
+      :~  [%group-path (path group-path.upd)]
+          [%app-name s+app-name.resource.upd]
+          [%app-path (path app-path.resource.upd)]
+      ==
+    [*@t *^json]
```
Comment (Fang-) [inline pkg/arvo/lib/metadata-json.hoon]: Can't you easily handle the `%associations` case with the `+assocations-to-json` from above? Seems weird to ever produce `{'': null}`. (Also this whole thing should probably be a `?-  -.upd`.)

File: pkg/arvo/app/chat-view.hoon [hunk 4]
```diff
@@ -249,6 +261,19 @@
       ==
     ~[(contact-view-poke [%create path ships])]
   ::
+  ++  create-metadata
+    |=  [group-path=path app-path=path]
+    ^-  (list card)
+    =/  =metadata
+      %*  .  *metadata
+        title         ''
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: I don't want to do that. I'd prefer to default to empty.

File: pkg/arvo/app/metadata-store.hoon [hunk 5]
```diff
@@ -1,3 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/metadata-store.hoon]: remvoed

File: pkg/arvo/lib/metadata-json.hoon (hunk 1 above)
Comment (loganallenc) [reply pkg/arvo/lib/metadata-json.hoon]: I'll make this a better data structure. Good critique.

File: pkg/arvo/lib/metadata-json.hoon (hunk 2 above)
Comment (loganallenc) [reply pkg/arvo/lib/metadata-json.hoon]: This is a concern for later. The frontend is used to dealing with renders that look like these, and I don't want to change them all wholesale right now.

File: pkg/arvo/app/chat-view.hoon [hunk 6]
```diff
@@ -199,39 +199,51 @@
...
     %-  zing
-    :~  :~  (chat-hook-poke [%remove path.act])
-            (permission-hook-poke [%remove path.act])
-            (chat-poke [%delete path.act])
+    :~  :~  (chat-hook-poke [%remove app-path.act])
+            (metadata-store-poke [%remove group-path [%chat app-path.act]])
+            (chat-poke [%delete app-path.act])
+        ==
+      ::
+        ?.  =(i.group-path '~')  ~
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: %chat-view is being used more or less as a library function. `+is-managed` is a good idea, and we should do it later.

File: pkg/arvo/app/chat-view.hoon [hunk 7]
```diff
@@ -52,8 +54,6 @@
       [%permission-group-hook-action permission-group-hook-action]
   ==
 --
-%-  agent:dbug
-%+  verb  |
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: Okay. Will put them back. Read the libraries and you're right that they don't get in the way of normal use.

File: pkg/arvo/lib/metadata-json.hoon (hunk 3 above)
Comment (loganallenc) [reply pkg/arvo/lib/metadata-json.hoon]: Yes, changed it.

File: pkg/arvo/lib/metadata-json.hoon (hunk 1 above)
Comment (loganallenc) [reply pkg/arvo/lib/metadata-json.hoon]: Fixed the structure to include the data in a separable format.

File: pkg/arvo/app/chat-view.hoon [hunk 8]
```diff
@@ -199,39 +199,51 @@
...
-        ?.  =(i.path.act '~')  ~
-        ~[(group-poke [%unbundle path.act])]
     ==
   ::
       %join
-    :~  (chat-hook-poke [%add-synced ship.act path.act ask-history.act])
-        (permission-hook-poke [%add-synced ship.act path.act])
+    ?>  ?=(^ app-path.act)
+    =/  group-path
+      ?:  =(i.app-path.act '~')  app-path.act
```
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: I think it's best to include the /~ prefixes with app-paths. It makes the join case significantly simpler to handle across all our applications.

[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: /~/~zod/path app-paths have a 1:1 correspondence to their unmanaged group.
/~zod/path app-paths have a many:1 correspondence to their managed group.

[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: For the unmanaged group case, you want to clean up and kill the group when the app-path associated with it is deleted. For the managed group, you don't want to clean up and kill the group when the app-path is deleted.

[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: For the join case, you don't want the user to have to pass in both the app-path and the group-path in order to join a particular chatroom associated with an unmanaged group, so you need a 1:1 correspondence between them.

For a managed group, you need to have *already* accepted the invite to the managed group to join the chat.

File: pkg/arvo/lib/metadata-json.hoon (hunk 2 above)
Comment (Fang-) [reply pkg/arvo/lib/metadata-json.hoon]: Dang, I should've caught this earlier then. But both this and the empty title thing signal that we're pushing a lot of things out to the frontend. Questionable imo, but a later discussion.

File: pkg/arvo/lib/metadata-json.hoon (hunk 1 above)
Comment (Fang-) [reply pkg/arvo/lib/metadata-json.hoon]: The "Even if ..." from above still applies, I'm pretty sure. Still sub-optimal, but functional enough?

File: pkg/arvo/app/chat-view.hoon (hunk 4 above)
[the remark this answers is not in the document]
Comment (ohAitch) [reply pkg/arvo/app/chat-view.hoon]: isn't title already going to bunt to `''` then?
```suggestion
```

[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: Removed this

File: pkg/arvo/app/chat-view.hoon (hunk 6 above)
[the remark this answers is not in the document]
Comment (loganallenc) [reply pkg/arvo/app/chat-view.hoon]: Added as a local helper.

**Comment by loganallenc** [comment]
@matildepark Would you mind doing a quick sanity check of the JS changes?

**Pull request merged**